### DISCUSSION TITLE
How to verify flashing

### DISCUSSION THREAD

**Discussion by CroftAr** [micropython/micropython#9654] [Hardware & Peripherals]
Is there a way to check that Micropython has been flashed correctly, both generally and for RP2040 in particular?  When flashing with 'esptool.py'  a '--verify' flag is available and the esp module provides 'esp.check_fw()' at run time, but I haven't found these or similar for non-ESP targets.

https://docs.micropython.org/en/latest/rp2/tutorial/intro.html
https://docs.micropython.org/en/latest/esp8266/tutorial/intro.html

**Comment by jimmo** [comment]
@CroftAr I think https://github.com/raspberrypi/picotool is the tool you're looking for.

**Reply by CroftAr** [reply]
Thanks!  I looked for a picotool .deb download but found only .rpm (in pkgs.org).  Is it wiser to get users to build their own, and why is that?

**Reply by jimmo** [reply]
Sorry, I guess that's a question for Raspberry Pi... It would be nice if the provided a .deb with each release (or better got it added to the distros).

**Comment by CroftAr** [comment]
Lastly is it possible to verify the f/w on a pyboard?  dfu-util can retrieve the board's f/w presumably in .bin format, and then...?

**Reply by jimmo** [reply]
Yep, `dfu-util -U` will let you extract retrieve the firmware (they call it "upload" which always seems back-to-front to me).

But this will be the whole flash image, so you'll need to extract out the relevant part to compare to the region that was flashed. Depending on what you started with (i.e. a .elf or .bin or .dfu) this will require different tools.

**Reply by dhylands** [reply]
There are a couple of gotchas. See this post: https://forum.micropython.org/viewtopic.php?f=3&t=5877#p33694 For whatever reason dfu-util won't let you cross segment boundaries. That particular post was talking about the filesystem, but the firmware is just at different addresses.
